sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/LockRenewSubscriber.java

```diff
@@ -0,0 +1,118 @@
+    private final Duration maxAutoLockRenewal;
+    private final LockContainer<LockRenewalOperation> messageLockContainer;
+    private final CoreSubscriber<? super ServiceBusReceivedMessage> actual;
+
+    private volatile Subscription subscription;
+    private static final AtomicReferenceFieldUpdater<LockRenewSubscriber, Subscription> UPSTREAM =
+        AtomicReferenceFieldUpdater.newUpdater(LockRenewSubscriber.class, Subscription.class,
+            "subscription");
+
+    LockRenewSubscriber(CoreSubscriber<? super ServiceBusReceivedMessage> actual, boolean autoLockRenewal,
+        Duration maxAutoLockRenewDuration, LockContainer<LockRenewalOperation> messageLockContainer,
+        Function<String, Mono<OffsetDateTime>> onRenewLock) {
+        this.onRenewLock = Objects.requireNonNull(onRenewLock, "'onRenewLock' cannot be null.");
+        this.actual = actual;
+        this.messageLockContainer = Objects.requireNonNull(messageLockContainer,
+            "'messageLockContainer' cannot be null.");
+
+        this.maxAutoLockRenewal = Objects.requireNonNull(maxAutoLockRenewDuration,
+            "'maxAutoLockRenewDuration' cannot be null.");
+        this.isAutoRenewLock = autoLockRenewal;
+
+    }
+    /**
+     * On an initial subscription, will take the first work item, and request that amount of work for it.
+     * @param subscription Subscription for upstream.
+     */
+    @Override
+    protected void hookOnSubscribe(Subscription subscription) {
+        Objects.requireNonNull(subscription, "'subscription' cannot be null.");
+        if (Operators.setOnce(UPSTREAM, this, subscription)) {
+            this.subscription = subscription;
+            actual.onSubscribe(subscription);
+        } else {
+            logger.error("Already subscribed once.");
+        }
+    }
+
+    /**
+     * When upstream has completed emitting messages.
+     */
+    @Override
+    public void hookOnComplete() {
+        logger.info("Upstream has completed.");
+        actual.onComplete();
+    }
+
+    @Override
+    protected void hookOnError(Throwable throwable) {
+        logger.error("Errors occurred upstream.");
+        actual.onError(throwable);
+        dispose();
+    }
+
+    @Override
+    protected void hookOnCancel() {
+        logger.error("Upstream cancelled.");
+        subscription.cancel();
+    }
+
+    @Override
+    protected void hookOnNext(ServiceBusReceivedMessage message) {
+        final String lockToken = message.getLockToken();
+        final OffsetDateTime lockedUntil = message.getLockedUntil();
+        final LockRenewalOperation renewOperation;
+
+        if (isAutoRenewLock && !Objects.isNull(lockedUntil) && !Objects.isNull(lockToken)) {
+
+            renewOperation = new LockRenewalOperation(lockToken, maxAutoLockRenewal, false, onRenewLock,
+                lockedUntil);
+            messageLockContainer.addOrUpdate(lockToken, OffsetDateTime.now().plus(maxAutoLockRenewal), renewOperation);
```

Comment:
What happens if this operation throws an exception? This can throw if the lockContainer is disposed.

Reply:
the error is thrown in this case.